File: meeting-scheduler/go/room.go
package main
type Room struct{
  name string
  calendar []Meeting
}



func (room *Room) book(startTime int64, endTime int64) (bool){
  //add validation
  isValidBooking, _:= room.isValidBooking(startTime, endTime)
  if isValidBooking == false {
    return false;
  }
  meeting := Meeting {
    startTime: startTime,
    endTime: endTime,
  }
  room.calendar = append(room.calendar, meeting)
  return true
}


func(room *Room) isValidBooking(startTime int64, endTime int64) (bool, error) {
  for _,v := range room.calendar {
    //check for overlap
    if (endTime >= v.startTime && endTime <= v.endTime) || (startTime >= v.startTime && startTime <= v.endTime) {
      return false , nil
    }
  }
  return true, nil;
}

File: meeting-scheduler/typescript/room.ts
import { Meeting } from "./meeting";

interface IRoom{
  book(startTime: number, endTime: number): boolean;
}

export class Room implements IRoom{
  public name: string;
  public  calendar : Meeting[];

  constructor(name: string) {
    this.name = name;
    this.calendar = [];
  }
   public book(startTime: number, endTime: number): boolean {
    //validation
    if(!this.isBookingValid(startTime, endTime)){
      return false;
    }
    let meeting = new Meeting(startTime, endTime);
    this.calendar.push(meeting);
    return true;
  }

  private isBookingValid(startTime: number, endTime: number): boolean{
    for(const slot of this.calendar){
      if((startTime >= slot.startTime && startTime <= slot.endTime) || (endTime >= slot.startTime && endTime <= slot.endTime)){
        return false;
      }
    }
    return true;
  }
}
